chartfx-chart/src/main/java/de/gsi/chart/renderer/spi/ContourDataSetRenderer.java

```diff
@@ -27,194 +47,125 @@
-                final double offset = (axisTransform.forward(z) - zMin) / (zMax - zMin);
-                data[indexYMax - 1 - yIndex][xIndex - indexXMin] = offset;
+
+        final int xSize = lCache.xSize;
+        final int ySize = lCache.ySize;
+        final double[][] data = new double[ySize][xSize];
+        for (int yIndex = 0; yIndex < ySize; yIndex++) {
+            for (int xIndex = 0; xIndex < xSize; xIndex++) {
+                final double offset = lCache.reduced[yIndex * xSize + xIndex];
+                data[ySize - 1 - yIndex][xIndex] = offset;
             }
         }
-        final MarchingSquares marchingSquares = new MarchingSquares();
-        try {
 
-            gc.save();
-            gc.scale(scaleX, scaleY);
+        // abort if min/max == 0 -> cannot compute contours
+        final double zRange = Math.abs(lCache.zMax - lCache.zMin);
+        if (zRange <= 0) {
+            return;
+        }
 
-            final GeneralPath[] isolines = marchingSquares.buildContours(data, levels);
+        final ColorGradient colorGradient = getColorGradient();
+        final MarchingSquares marchingSquares = new MarchingSquares();
+        final double scaleX = lCache.xDataPixelRange / xSize;
+        final double scaleY = lCache.yDataPixelRange / ySize;
+        gc.save();
+        gc.translate(lCache.xDataPixelMin, lCache.yDataPixelMin);
+        gc.scale(scaleX, scaleY);
+        final GeneralPath[] isolines;
+        try {
+            isolines = marchingSquares.buildContours(data, levels);
             int levelCount = 0;
             for (final GeneralPath path : isolines) {
                 if (path.size() > getMaxContourSegments()) {
                     levelCount++;
                     continue;
                 }
-                final Color color = lCache.zInverted ? getColor(1 - levels[levelCount++])
-                        : getColor(levels[levelCount++]);
+                final Color color = lCache.zInverted ? colorGradient.getColor(1 - levels[levelCount++])
+                                                     : colorGradient.getColor(levels[levelCount++]);
                 gc.setStroke(color);
                 gc.setLineDashes(1.0);
                 gc.setMiterLimit(10);
                 gc.setFill(color);
                 gc.setLineWidth(0.5);
                 path.draw(gc);
             }
-            gc.restore();
-
         } catch (InterruptedException | ExecutionException e) {
-            // TODO Auto-generated catch block
-            e.printStackTrace();
+            if (LOGGER.isErrorEnabled()) {
+                LOGGER.atError().setCause(e).log("marchingSquares algorithm");
+            }
+        } finally {
+            gc.restore();
         }
     }
 
-    private void drawContourFast(final GraphicsContext gc, final AxisTransform axisTransform, final Cache lCache) {
+    private void drawContourFast(final GraphicsContext gc, final AxisTransform axisTransform,
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [The method 'drawContourFast(GraphicsContext, AxisTransform, DataSetCache)' has an NPath complexity of 504](https://app.codacy.com/manual/GSI/chart-fx/pullRequest?prid=4880492)